src/agent.rs

```diff
@@ -43,79 +43,99 @@ impl Decoder for MessageCodec {
     }
 }
 
-impl Encoder for MessageCodec {
-    type Item = Message;
+impl Encoder<Message> for MessageCodec {
     type Error = AgentError;
 
-    fn encode(&mut self, item: Self::Item, dst: &mut BytesMut) -> Result<(), Self::Error> {
+    fn encode(&mut self, item: Message, dst: &mut BytesMut) -> Result<(), Self::Error> {
         let bytes = to_bytes(&to_bytes(&item)?)?;
-        dst.put(bytes);
+        dst.put(&*bytes);
         Ok(())
     }
 }
 
-macro_rules! handle_clients {
-    ($self:ident, $socket:ident) => {{
-        info!("Listening; socket = {:?}", $socket);
-        let arc_self = Arc::new($self);
-        $socket
-            .incoming()
-            .map_err(|e| error!("Failed to accept socket; error = {:?}", e))
-            .for_each(move |socket| {
-                let (write, read) = Framed::new(socket, MessageCodec).split();
-                let arc_self = arc_self.clone();
-                let connection = write
-                    .send_all(read.and_then(move |message| {
-                        arc_self.handle_async(message).map_err(|e| {
-                            error!("Error handling message; error = {:?}", e);
-                            AgentError::User
-                        })
-                    }))
-                    .map(|_| ())
-                    .map_err(|e| error!("Error while handling message; error = {:?}", e));
-                tokio::spawn(connection)
-            })
-            .map_err(|e| e.into())
-    }};
+struct Session<A, S> {
+    agent: Arc<A>,
+    adapter: Framed<S, MessageCodec>,
 }
 
-pub trait Agent: 'static + Sync + Send + Sized {
-    type Error: Debug + Send + Sync;
-
-    fn handle(&self, message: Message) -> Result<Message, Self::Error>;
-
-    fn handle_async(
-        &self,
-        message: Message,
-    ) -> Box<dyn Future<Item = Message, Error = Self::Error> + Send + Sync> {
-        Box::new(FutureResult::from(self.handle(message)))
+impl<A, S> Session<A, S>
+where
+    A: Agent,
+    S: AsyncRead + AsyncWrite + Unpin,
+{
+    fn new(agent: Arc<A>, socket: S) -> Self {
+        let adapter = Framed::new(socket, MessageCodec);
+        Self { agent, adapter }
     }
 
-    #[allow(clippy::unit_arg)]
-    fn run_listener(self, socket: UnixListener) -> Result<(), Box<dyn Error + Send + Sync>> {
-        Ok(tokio::run(handle_clients!(self, socket)))
+    async fn handle_socket(&mut self) -> Result<(), AgentError> {
+        loop {
+            match self.adapter.try_next().await? {
```

Comment:
:thinking: What do you think about converting this `match` to an `if let Some(incoming_message) = ...`? (since it's just two arms and one of them is a `None`)